lib/vector-config/src/schema/parser/query.rs

```diff
@@ -0,0 +1,435 @@
+use std::{fs::File, io::BufReader, path::Path};
+
+use once_cell::sync::OnceCell;
+use serde_json::Value;
+use snafu::Snafu;
+use vector_config_common::{
+    attributes::CustomAttribute,
+    constants,
+    schema::{InstanceType, RootSchema, Schema, SchemaObject, SingleOrVec},
+};
+
+#[derive(Debug, Snafu)]
+#[snafu(module, context(suffix(false)))]
+pub enum QueryError {
+    #[snafu(display("I/O error during opening schema: {source}"), context(false))]
+    Io { source: std::io::Error },
+
+    #[snafu(display("deserialization failed: {source}"), context(false))]
+    Deserialization { source: serde_json::Error },
+
+    #[snafu(display("no schemas matched the query"))]
+    NoMatches,
+
+    #[snafu(display("multiple schemas matched the query ({len})"))]
+    MultipleMatches { len: usize },
+
+    #[snafu(display("found matching attribute but was not a flag"))]
+    AttributeNotFlag,
+
+    #[snafu(display(
+        "found matching attribute but expected single value; multiple values present"
+    ))]
+    AttributeMultipleValues,
+}
+
+pub struct SchemaQuerier {
+    schema: RootSchema,
+}
+
+impl SchemaQuerier {
+    /// Creates a `SchemaQuerier` based on the schema file located at `schema_path`.
+    ///
+    /// # Errors
+    ///
+    /// If no file exists at the given schema path, or there is an I/O error during loading the file
+    /// (permissions, etc), then an error variant will be returned.
+    ///
+    /// If the file does not contain valid JSON, or cannot be deserialized as a schema, then an
+    /// error variant will be returned.
+    pub fn from_schema<P: AsRef<Path>>(schema_path: P) -> Result<Self, QueryError> {
+        let reader = File::open(schema_path).map(BufReader::new)?;
+        let schema = serde_json::from_reader(reader)?;
+
+        Ok(Self { schema })
+    }
+
+    pub fn query(&self) -> SchemaQueryBuilder<'_> {
+        SchemaQueryBuilder::from_schema(&self.schema)
+    }
+}
+
+/// A query builder for querying against a root schema.
+///
+/// All constraints are applied in a boolean AND fashion.
+pub struct SchemaQueryBuilder<'a> {
+    schema: &'a RootSchema,
+    attributes: Vec<CustomAttribute>,
+}
+
+impl<'a> SchemaQueryBuilder<'a> {
+    fn from_schema(schema: &'a RootSchema) -> Self {
+        Self {
+            schema,
+            attributes: Vec::new(),
+        }
+    }
+
+    /// Adds a constraint on the given custom attribute key/value.
+    ///
+    /// Can be used multiple times to match schemas against multiple attributes.
+    ///
+    /// Custom attributes are strongly matched: a flag attribute can only match a flag attribute,
+    /// not a key/value attribute, and vise versa. For key/value attributes where the attribute in
```

Comment:
```suggestion
    /// not a key/value attribute, and vice versa. For key/value attributes where the attribute in
```

Reply:
Cleaned up in 03262ef97... along with all the other incorrect usages I could find. :D